### PR TITLE
Use a service user to not run container process as root

### PR DESCRIPTION
**Pull request by chr4**
What: Create a `storj` user and group in Dockerfile, so the process doesn't run as root

Why: For security. Also, it's a The [official best practise](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user)

Please describe the tests: None added so far. Required?
 
Please describe the performance impact: Should be none.

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: cmd/storagenode/Dockerfile [hunk 1]
```diff
@@ -10,6 +10,8 @@ COPY resources/certs.pem /etc/ssl/certs/ca-certificates.crt
 COPY release/${TAG}/storagenode_linux_${GOARCH:-amd64} /app/storagenode
 COPY cmd/storagenode/entrypoint /entrypoint
 COPY cmd/storagenode/dashboard.sh /app/dashboard.sh
+RUN addgroup storj && adduser -D -G storj storj && chown -R storj:storj /app
```
Comment (brimstone) [inline cmd/storagenode/Dockerfile]: Does the directory of binaries need to be owned by the storj user? The user doesn't change the binaries at all and should have access already to run them.

Comment (chr4) [reply cmd/storagenode/Dockerfile]: Good question. I'm not sure. In case this is read only anyway, the `chown` should be skipped.

**Comment by chr4** [comment]
I agree, that there should be a smooth upgrade path for volumes. I think this is out of the scope of `Dockerfile`, as the volum (afaik) can't be mounted. 

Where would such a change be placed? Maybe in `docker-compose.yml` or even the `Makefile`? For this, it might be helpful to hardcode the `storj` user's UID.

**Comment by utdrmac** [comment]
Can this be merged please?

**Comment by ionphractal** [comment]
> I agree, that there should be a smooth upgrade path for volumes. I think this is out of the scope of `Dockerfile`, as the volum (afaik) can't be mounted.
> 
> Where would such a change be placed? Maybe in `docker-compose.yml` or even the `Makefile`? For this, it might be helpful to hardcode the `storj` user's UID.

The challenge here is that if you start the container without root, you will not be able to modify the data directory permissions. AFAIK there is only one way, though in a two-step upgrade it could be possible to give up on root entirely.

In the first step you would basically run the container still as root, but drop privileges as soon as you run the node executable. This would be done in the entrypoint script (https://github.com/storj/storj/blob/5395ff5fe6c019e2773450900631f1e6c6353046/cmd/storagenode/entrypoint#L22). You need a helper like runit's chpst in the container, which is imho the best choice for Alpine containers (http://jdebp.eu./FGA/dont-abuse-su-for-dropping-privileges.html , https://unix.stackexchange.com/questions/132663/how-do-i-drop-root-privileges-in-shell-scripts#132676).

If you really want to drop all root usage, you can change the Dockerfile to run as the created user *in the next version*. However, you would have to force your users to obey update order... which is hard if not impossible.

Btw you may already drop if using docker-compose for example using the `guest` user:
* add `user: guest` to your container `docker-compose.yml`
* `docker-compose stop` and `rm` on it
* modify all your files to `405:100` in the container
* `docker-compose up` again

**Comment by chr4** [comment]
Another alternative would be to use [usernd-remap](https://docs.docker.com/engine/security/userns-remap/) to map the `root` user in the container to an unprivileged host user. 

The problems with automatically `chown`ing files would be the same, but this would allow more tech-savy users to run the container in a less privileged way.

**Comment by stefanbenten** [comment]
@chr4 @ionphractal Would anyone of you be willing to get this further along?
This PR is stale for quite some time now..

**Comment by ionphractal** [comment]
> @chr4 @ionphractal Would anyone of you be willing to get this further along?
> This PR is stale for quite some time now..

@stefanbenten For my part I'm using the docker-compose.yml mechanism I provided. 

To make something like this default, I don't see a feasible way to address this without the help of manual migration instructions.

- usernd-remap requires manual configuration of the docker daemon, which I don't want to be forced to do as a user. Besides, it makes setup more complicated as it affects all other docker files/containers/volumes on the system. It would've been great if this was docker default to begin with, but it isn't.
- I'm against always using chown at startup in the container because we're talking loads of little files (e.g. I have > 270 000 files for 500GB) which will delay the node startup immensely, depending on the underlying storage layer. On the magnetic disks in the system I use it, said volume currently takes >9 minutes. There could be a run-once detection though, but I wonder how good this works in the end.

But I think an official standpoint needs to be taken first: "Should the default user be "not root"? If yes, what name/UID/GUID should it be?"

If that is clear, then I suggest to announce the change to the users and that the startup in one of the next versions will be longer, depending on their volume etc.

**Comment by chr4** [comment]
Signed CLA.

**Comment by Erikvv** [comment]
What currently prevents the user from using `docker run --user $UID` ?

I see this mainly as a runtime decision. 

(If the username exists in the container it might be necessary to start the container as root, then pass the UID via an environment variable. A bit like I did in https://github.com/storj-thirdparty/uplink-php/blob/main/Jenkinsfile#L43)

File: cmd/storagenode/Dockerfile [hunk 2]
```diff
@@ -14,6 +14,8 @@ COPY --from=ca-cert /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certifi
...
```
[the remark this answers is not in the document]
Comment (chr4) [reply cmd/storagenode/Dockerfile]: True, changed.

[the remark this answers is not in the document]
Comment (chr4) [reply cmd/storagenode/Dockerfile]: Changed to `adduser`, which only supports the short args unfortunately.

[the remark this answers is not in the document]
Comment (chr4) [reply cmd/storagenode/Dockerfile]: Changed to use long options for `chown` and `adduser`

**Comment by Erikvv** [comment]
I want to like this change, because Storj by default shouldn't take more permissions than necessary. Both on the user system and inside the container.

But the problem is like @brimstone said:

> But to approve this, it needs to handle the ownership of the data directory in the case of:
> 
> 1. Fresh host volume. The default is the root user owns the directory.

I think assigning a random user id inside the container is a not a good starting point. The only way we are going to make that work is to make the user create (or at least chown) this user id on his system. This is an extra burden and the user might already be taken for other purposes also.

To remove this action we can move the logic from the Dockerfile to the entrypoint to `chown` the volume there. I think that's still working against the system. Having an unknown user id on the system might also have some side-effects.

**I propose to abandon this change, and update our documentation** at https://docs.storj.io/node/setup/cli/storage-node

1. Instruct the user to create the mount point beforehand: `mkdir -p <storage-dir>`
2. Add the flag `--user <user>` to the `docker run` commands. Alternatively  `--user $(id -u)`.

What do you think? can we a agree on this?

**Comment by chr4** [comment]
I think `--user` covers the main point of this pull request, being that storj should be run as root. 
Things like setting/ allowing `userns` might offer some more protection - but this might already work with some custom `chown` for the data dir.

I think encouraging users to use `--user $UID` is a good compromise!

**Comment by Erikvv** [comment]
I have asked the documentation people to pick this up. https://github.com/storj/storj/issues/4458

@chr4 Sorry for asking you to do the changes, and thanks for bringing this to our attention.

**Pull request closed without merging**